lib/db/sql/mapper.php

```diff
@@ -248,30 +248,32 @@ function($str) use($db) {
 		if (preg_match('/mssql|sqlsrv|odbc/', $this->engine) &&
 			($options['limit'] || $options['offset'])) {
 			$pkeys=[];
-			foreach ($this->fields as $key=>$field)
-				if ($field['pkey'])
-					$pkeys[]=$key;
+			if (!$options['order']) {
+				foreach ($this->fields as $key=>$field)
+					if ($field['pkey'])
+						$pkeys[]=$key;
+				$ord=' ORDER BY '.$db->quotekey($pkeys[0]);
+			}
 			$ofs=$options['offset']?(int)$options['offset']:0;
 			$lmt=$options['limit']?(int)$options['limit']:0;
 			if (strncmp($db->version(),'11',2)>=0) {
 				// SQL Server 2012
-				if (!$options['order'])
-					$sql.=' ORDER BY '.$db->quotekey($pkeys[0]);
-				$sql.=' OFFSET '.$ofs.' ROWS';
+				$sql.=$ord.' OFFSET '.$ofs.' ROWS';
 				if ($lmt)
 					$sql.=' FETCH NEXT '.$lmt.' ROWS ONLY';
 			}
 			else {
 				// SQL Server 2008
-				$sql=str_replace('SELECT',
```

Comment:
replacing all the SELECT's breaks subselects in the adhoc fields and filters. We really just want to replace the first one.